# Data cleansing

In this notebook we will cover the process of cleaning the data in different stages: 

1. We will load the raw dataset of FIFA 18 players and check some of its properties.

2. We will filter the dataset to only use rows and columns of our interest.

3. We will transform required fields in order to perform a cleaner analysis.

4. We will deal with missing values, droping or filling them, whenever necessary.

5. We will save the processed dataset in a new folder.

That way we will simulate an ETL process: Extract (step 1), Transform (steps 2, 3 and 4) and Load (step 5).

We will be using functions defined in the script '../src/transfromations.py'. Also, alongside this notebook, we will be checking the effects of our transformations onto the dataset.

## Step 1. Loading raw data.

First step of our cleaning process is extracting the data from its source and import the necessary modules that we are going to use alongside this phase.

In [1]:
# These lines allow to check if transformation.py has changed before execute another cell
%load_ext autoreload
%autoreload 2

import sys
from pathlib import Path

src_path = Path("..") / "src"
sys.path.append(str(src_path))

In [2]:
import pandas as pd
import transformations as tr

We read the csv file and show a few lines, info and description.

In [ ]:
data = pd.read_csv('../data/raw/fifa.csv')

data.head()

Unnamed: 0      ID               Name  Age  \
0           0  158023           L. Messi   31   
1           1   20801  Cristiano Ronaldo   33   
2           2  190871          Neymar Jr   26   
3           3  193080             De Gea   27   
4           4  192985       K. De Bruyne   27   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1   https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2  https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3  https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4  https://cdn.sofifa.org/players/4/19/192985.png     Belgium   

                                  Flag  Overall  Potential  \
0  https://cdn.sofifa.org/flags/52.png       94         94   
1  https://cdn.sofifa.org/flags/38.png       94         94   
2  https://cdn.sofifa.org/flags/54.png       92         93   
3  https://cdn.sofifa.org/flags/45.png       91         93   
4   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club  ... Composure Marking StandingTackle  SlidingTackle  \
0         FC Barcelona  ...      96.0    33.0           28.0           26.0   
1             Juventus  ...      95.0    28.0           31.0           23.0   
2  Paris Saint-Germain  ...      94.0    27.0           24.0           33.0   
3    Manchester United  ...      68.0    15.0           21.0           13.0   
4      Manchester City  ...      88.0    68.0           58.0           51.0   

  GKDiving  GKHandling  GKKicking  GKPositioning GKReflexes Release Clause  
0      6.0        11.0       15.0           14.0        8.0        €226.5M  
1      7.0        11.0       15.0           14.0       11.0        €127.1M  
2      9.0         9.0       15.0           15.0       11.0        €228.1M  
3     90.0        85.0       87.0           88.0       94.0        €138.6M  
4     15.0        13.0        5.0           10.0       13.0        €196.4M  

[5 rows x 89 columns]

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18207 entries, 0 to 18206
Data columns (total 89 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Unnamed: 0                18207 non-null  int64  
 1   ID                        18207 non-null  int64  
 2   Name                      18207 non-null  object 
 3   Age                       18207 non-null  int64  
 4   Photo                     18207 non-null  object 
 5   Nationality               18207 non-null  object 
 6   Flag                      18207 non-null  object 
 7   Overall                   18207 non-null  int64  
 8   Potential                 18207 non-null  int64  
 9   Club                      17966 non-null  object 
 10  Club Logo                 18207 non-null  object 
 11  Value                     18207 non-null  object 
 12  Wage                      18207 non-null  object 
 13  Special                   18207 non-null  int64  
 14  Prefer

In [ ]:
data.describe()

Unnamed: 0             ID           Age       Overall     Potential  \
count  18207.000000   18207.000000  18207.000000  18207.000000  18207.000000   
mean    9103.000000  214298.338606     25.122206     66.238699     71.307299   
std     5256.052511   29965.244204      4.669943      6.908930      6.136496   
min        0.000000      16.000000     16.000000     46.000000     48.000000   
25%     4551.500000  200315.500000     21.000000     62.000000     67.000000   
50%     9103.000000  221759.000000     25.000000     66.000000     71.000000   
75%    13654.500000  236529.500000     28.000000     71.000000     75.000000   
max    18206.000000  246620.000000     45.000000     94.000000     95.000000   

            Special  International Reputation     Weak Foot   Skill Moves  \
count  18207.000000              18159.000000  18159.000000  18159.000000   
mean    1597.809908                  1.113222      2.947299      2.361308   
std      272.586016                  0.394031      0.660456      0.756164   
min      731.000000                  1.000000      1.000000      1.000000   
25%     1457.000000                  1.000000      3.000000      2.000000   
50%     1635.000000                  1.000000      3.000000      2.000000   
75%     1787.000000                  1.000000      3.000000      3.000000   
max     2346.000000                  5.000000      5.000000      5.000000   

       Jersey Number  ...     Penalties     Composure       Marking  \
count   18147.000000  ...  18159.000000  18159.000000  18159.000000   
mean       19.546096  ...     48.548598     58.648274     47.281623   
std        15.947765  ...     15.704053     11.436133     19.904397   
min         1.000000  ...      5.000000      3.000000      3.000000   
25%         8.000000  ...     39.000000     51.000000     30.000000   
50%        17.000000  ...     49.000000     60.000000     53.000000   
75%        26.000000  ...     60.000000     67.000000     64.000000   
max        99.000000  ...     92.000000     96.000000     94.000000   

       StandingTackle  SlidingTackle      GKDiving    GKHandling  \
count    18159.000000   18159.000000  18159.000000  18159.000000   
mean        47.697836      45.661435     16.616223     16.391596   
std         21.664004      21.289135     17.695349     16.906900   
min          2.000000       3.000000      1.000000      1.000000   
25%         27.000000      24.000000      8.000000      8.000000   
50%         55.000000      52.000000     11.000000     11.000000   
75%         66.000000      64.000000     14.000000     14.000000   
max         93.000000      91.000000     90.000000     92.000000   

          GKKicking  GKPositioning    GKReflexes  
count  18159.000000   18159.000000  18159.000000  
mean      16.232061      16.388898     16.710887  
std       16.502864      17.034669     17.955119  
min        1.000000       1.000000      1.000000  
25%        8.000000       8.000000      8.000000  
50%       11.000000      11.000000     11.000000  
75%       14.000000      14.000000     14.000000  
max       91.000000      90.000000     94.000000  

[8 rows x 44 columns]

We are not going to comment anything about the dataset, as it is already commented in the first notebook '01.exploration.ipynb'.

## Step 2. Filtering the dataset.

Now we want to reduce our dataset in two different ways:

1. Selecting only relevant characteristic obout the players (column filtering).

2. Selecting only globally relevant players (rows filtering).

Both of them are implemented in the transformation module function, 'filter'.

From now on, our dataset will be referred as 'data'.

In [ ]:
data_filtered = tr.filter(data)

data_filtered.head()

Name  Age Nationality  Overall  Potential  \
0           L. Messi   31   Argentina       94         94   
1  Cristiano Ronaldo   33    Portugal       94         94   
2          Neymar Jr   26      Brazil       92         93   
3             De Gea   27       Spain       91         93   
4       K. De Bruyne   27     Belgium       91         92   

   International Reputation                 Club   Wage Preferred Foot  \
0                       5.0         FC Barcelona  €565K           Left   
1                       5.0             Juventus  €405K          Right   
2                       5.0  Paris Saint-Germain  €290K          Right   
3                       4.0    Manchester United  €260K          Right   
4                       4.0      Manchester City  €355K          Right   

   Weak Foot Position  Jersey Number Height  Weight        Joined  \
0        4.0       RF           10.0    5'7  159lbs   Jul 1, 2004   
1        4.0       ST            7.0    6'2  183lbs  Jul 10, 2018   
2        5.0       LW           10.0    5'9  150lbs   Aug 3, 2017   
3        3.0       GK            1.0    6'4  168lbs   Jul 1, 2011   
4        5.0      RCM            7.0   5'11  154lbs  Aug 30, 2015   

  Contract Valid Until Release Clause  
0                 2021        €226.5M  
1                 2022        €127.1M  
2                 2022        €228.1M  
3                 2020        €138.6M  
4                 2023        €196.4M

Lets also look at some info and description about the new dataset.

In [ ]:
data_filtered.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1627 entries, 0 to 17466
Data columns (total 17 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Name                      1627 non-null   object 
 1   Age                       1627 non-null   int64  
 2   Nationality               1627 non-null   object 
 3   Overall                   1627 non-null   int64  
 4   Potential                 1627 non-null   int64  
 5   International Reputation  1627 non-null   float64
 6   Club                      1613 non-null   object 
 7   Wage                      1627 non-null   object 
 8   Preferred Foot            1627 non-null   object 
 9   Weak Foot                 1627 non-null   float64
 10  Position                  1627 non-null   object 
 11  Jersey Number             1627 non-null   float64
 12  Height                    1627 non-null   object 
 13  Weight                    1627 non-null   object 
 14  Joined      

In [ ]:
data_filtered.describe()

Age      Overall    Potential  International Reputation  \
count  1627.000000  1627.000000  1627.000000               1627.000000   
mean     29.186232    77.049785    78.107560                  2.263675   
std       4.001274     4.968967     5.734285                  0.528338   
min      18.000000    54.000000    54.000000                  2.000000   
25%      26.000000    74.000000    74.000000                  2.000000   
50%      29.000000    77.000000    77.000000                  2.000000   
75%      32.000000    80.000000    82.000000                  2.000000   
max      45.000000    94.000000    95.000000                  5.000000   

         Weak Foot  Jersey Number  
count  1627.000000    1627.000000  
mean      3.210203      15.791026  
std       0.726459      13.388989  
min       1.000000       1.000000  
25%       3.000000       7.000000  
50%       3.000000      13.000000  
75%       4.000000      21.000000  
max       5.000000      99.000000

As we can see, we have reduced our dataset to 1627 players, with no missing values in numerical columns. We also have reduced missing values in the rest of the fields.

## Step 3. Transformations.

In this step we want to transform every column needed to perform a clean analysis and be able to deal with missing values. More accurately, we want to:

1. Convert 'Height' and 'Weight' columns to the International System units, and storing only the quantity (units will be specified in the column header).

2. Convert 'Wage' and 'Release Clause' columns into numerical data (units will be specified in the column header).

3. Control that 'Joined' and 'Contract Valid Until' store valid and unified formats.

4. Create tactical groups for the 'Position' column.

5. Make sure that the rest of numeric columns are numerical and not strings (age, overall, potential, international reputation, weak foot, and jersey number). This step will be done after dealing with missing values.

First four steps are implemented by transformations function 'data_adjustment', which calls other transformation functions for each step.

In [ ]:
data_transformed = tr.data_adjustment(data_filtered)

data_transformed.head()

Name  Age Nationality  Overall  Potential  \
0           L. Messi   31   Argentina       94         94   
1  Cristiano Ronaldo   33    Portugal       94         94   
2          Neymar Jr   26      Brazil       92         93   
3             De Gea   27       Spain       91         93   
4       K. De Bruyne   27     Belgium       91         92   

   International Reputation                 Club      Wage Preferred Foot  \
0                       5.0         FC Barcelona  565000.0           Left   
1                       5.0             Juventus  405000.0          Right   
2                       5.0  Paris Saint-Germain  290000.0          Right   
3                       4.0    Manchester United  260000.0          Right   
4                       4.0      Manchester City  355000.0          Right   

   Weak Foot Position Position_Group  Jersey Number  Height  Weight  \
0        4.0       RF            ATT           10.0  170.18   72.12   
1        4.0       ST            ATT            7.0  187.96   83.01   
2        5.0       LW            ATT           10.0  175.26   68.04   
3        3.0       GK             GK            1.0  193.04   76.20   
4        5.0      RCM            MDF            7.0  180.34   69.85   

      Joined Contract Valid Until  Release Clause  
0 2004-07-01           2021-12-31     226500000.0  
1 2018-07-10           2022-12-31     127100000.0  
2 2017-08-03           2022-12-31     228100000.0  
3 2011-07-01           2020-12-31     138600000.0  
4 2015-08-30           2023-12-31     196400000.0

In [ ]:
data_transformed.describe()

Age      Overall    Potential  International Reputation  \
count  1627.000000  1627.000000  1627.000000               1627.000000   
mean     29.186232    77.049785    78.107560                  2.263675   
min      18.000000    54.000000    54.000000                  2.000000   
25%      26.000000    74.000000    74.000000                  2.000000   
50%      29.000000    77.000000    77.000000                  2.000000   
75%      32.000000    80.000000    82.000000                  2.000000   
max      45.000000    94.000000    95.000000                  5.000000   
std       4.001274     4.968967     5.734285                  0.528338   

                Wage    Weak Foot  Jersey Number       Height       Weight  \
count    1627.000000  1627.000000    1627.000000  1627.000000  1627.000000   
mean    46958.205286     3.210203      15.791026   182.019803    77.393301   
min         0.000000     1.000000       1.000000   157.480000    55.790000   
25%     15000.000000     3.000000       7.000000   177.800000    72.120000   
50%     30000.000000     3.000000      13.000000   182.880000    77.110000   
75%     56000.000000     4.000000      21.000000   187.960000    82.100000   
max    565000.000000     5.000000      99.000000   200.660000   101.150000   
std     55157.145278     0.726459      13.388989     6.650292     7.176719   

                              Joined           Contract Valid Until  \
count                           1506                           1613   
mean   2016-01-30 09:24:08.605577728  2021-05-18 16:02:22.839429632   
min              1991-06-01 00:00:00            2018-12-31 00:00:00   
25%              2015-02-02 00:00:00            2019-12-31 00:00:00   
50%              2017-01-17 00:00:00            2020-12-31 00:00:00   
75%              2018-01-30 00:00:00            2021-12-31 00:00:00   
max              2018-11-23 00:00:00            2026-12-31 00:00:00   
std                              NaN                            NaN   

       Release Clause  
count    1.627000e+03  
mean     2.097014e+07  
min      0.000000e+00  
25%      3.850000e+06  
50%      1.280000e+07  
75%      2.665000e+07  
max      2.281000e+08  
std      2.695895e+07

In [ ]:
data_transformed.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1627 entries, 0 to 17466
Data columns (total 18 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   Name                      1627 non-null   object        
 1   Age                       1627 non-null   int64         
 2   Nationality               1627 non-null   object        
 3   Overall                   1627 non-null   int64         
 4   Potential                 1627 non-null   int64         
 5   International Reputation  1627 non-null   float64       
 6   Club                      1613 non-null   object        
 7   Wage                      1627 non-null   float64       
 8   Preferred Foot            1627 non-null   object        
 9   Weak Foot                 1627 non-null   float64       
 10  Position                  1627 non-null   object        
 11  Position_Group            1627 non-null   object        
 12  Jersey Number           

## Step 4. Dealing with missing values.

Despite the fact that we have significantly reduce the number of missing values in each column, we still nead to deal with them. We have implemented a function that does so name missing_treatment for every column presented in the dataset. It performs the following actions depending on the column:

* **Drop rows with missing values**: Name, Overall, Potential, International Reputation and Position.

* **Age:** not missing values in our dataset. Replace missing malues with its mean.

* **Nationality:** not missing values in our dataset. Replace missing malues with 'Unknown'.

* **Club:** replace missing values with 'Free Agent'. 

* **Wage:** not missing values in our dataset. Replace missing malues with 0.

* **Preferred Foot:** not missing values in our dataset. Replace missing values with 'Right'. 

* **Weak Foot:** not missing values in our dataset. Replace missing values with the minimum value in the dataset.

* **Jersey Number:** not missing values in our dataset. Replace missing values with 0.

* **Height:** not missing values in our dataset. Replace missing malues with its mean.

* **Weight:** not missing values in our dataset. Replace missing malues with its mean.

* **Joined:** replace missing values with 'Jan 31, 2019' (virtually not joined yet, a free agent).

* **Contract Valid Until:** replace missing values with 'Dec 31, 2018' (current year).

* **Release Clause:** replace missing values with 0.

In [ ]:
data_completed = tr.missing_treatment(data_transformed)

data_completed.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1627 entries, 0 to 17466
Data columns (total 18 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   Name                      1627 non-null   object        
 1   Age                       1627 non-null   int64         
 2   Nationality               1627 non-null   object        
 3   Overall                   1627 non-null   int64         
 4   Potential                 1627 non-null   int64         
 5   International Reputation  1627 non-null   float64       
 6   Club                      1627 non-null   object        
 7   Wage                      1627 non-null   float64       
 8   Preferred Foot            1627 non-null   object        
 9   Weak Foot                 1627 non-null   float64       
 10  Position                  1627 non-null   object        
 11  Position_Group            1627 non-null   object        
 12  Jersey Number           

## Step 5. Final details

Now that we have filled all of the missing values, we only have to do the last few changes in order to improve our dataset presentation. These little changes are:

1. There are some float values (International Reputation, Weak Foot and Jersey Number) which should be ints. We will cast these columns into integers using the transformation function 'casting'.

2. We will scale currency columns making them more readable. 'Wage' column will be expressed in thousands of euros, while 'Release Clause' will be expressed in millions of euros. In addition, we will rename these columns to point out its scale (the same with 'Height' and Weight').

In [ ]:
data_casted = tr.cast_to_integer(data_completed)
data_renamed = tr.scale_rename(data_casted)

And now we show the final DataFrame and some information about it.

In [ ]:
data_renamed.head()

Name  Age Nationality  Overall  Potential  \
0           L. Messi   31   Argentina       94         94   
1  Cristiano Ronaldo   33    Portugal       94         94   
2          Neymar Jr   26      Brazil       92         93   
3             De Gea   27       Spain       91         93   
4       K. De Bruyne   27     Belgium       91         92   

   International Reputation                 Club  Wage (K_EUR) Preferred Foot  \
0                         5         FC Barcelona         565.0           Left   
1                         5             Juventus         405.0          Right   
2                         5  Paris Saint-Germain         290.0          Right   
3                         4    Manchester United         260.0          Right   
4                         4      Manchester City         355.0          Right   

   Weak Foot Position Position_Group  Jersey Number  Height_cm  Weight_kg  \
0          4       RF            ATT             10     170.18      72.12   
1          4       ST            ATT              7     187.96      83.01   
2          5       LW            ATT             10     175.26      68.04   
3          3       GK             GK              1     193.04      76.20   
4          5      RCM            MDF              7     180.34      69.85   

      Joined Contract Valid Until  Release Clause (M_EUR)  
0 2004-07-01           2021-12-31                   226.5  
1 2018-07-10           2022-12-31                   127.1  
2 2017-08-03           2022-12-31                   228.1  
3 2011-07-01           2020-12-31                   138.6  
4 2015-08-30           2023-12-31                   196.4

In [ ]:
data_renamed.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1627 entries, 0 to 17466
Data columns (total 18 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   Name                      1627 non-null   object        
 1   Age                       1627 non-null   int64         
 2   Nationality               1627 non-null   object        
 3   Overall                   1627 non-null   int64         
 4   Potential                 1627 non-null   int64         
 5   International Reputation  1627 non-null   int64         
 6   Club                      1627 non-null   object        
 7   Wage (K_EUR)              1627 non-null   float64       
 8   Preferred Foot            1627 non-null   object        
 9   Weak Foot                 1627 non-null   int64         
 10  Position                  1627 non-null   object        
 11  Position_Group            1627 non-null   object        
 12  Jersey Number           

In [ ]:
data_renamed.describe()

Age      Overall    Potential  International Reputation  \
count  1627.000000  1627.000000  1627.000000               1627.000000   
mean     29.186232    77.049785    78.107560                  2.263675   
min      18.000000    54.000000    54.000000                  2.000000   
25%      26.000000    74.000000    74.000000                  2.000000   
50%      29.000000    77.000000    77.000000                  2.000000   
75%      32.000000    80.000000    82.000000                  2.000000   
max      45.000000    94.000000    95.000000                  5.000000   
std       4.001274     4.968967     5.734285                  0.528338   

       Wage (K_EUR)    Weak Foot  Jersey Number    Height_cm    Weight_kg  \
count   1627.000000  1627.000000    1627.000000  1627.000000  1627.000000   
mean      46.958205     3.210203      15.791026   182.019803    77.393301   
min        0.000000     1.000000       1.000000   157.480000    55.790000   
25%       15.000000     3.000000       7.000000   177.800000    72.120000   
50%       30.000000     3.000000      13.000000   182.880000    77.110000   
75%       56.000000     4.000000      21.000000   187.960000    82.100000   
max      565.000000     5.000000      99.000000   200.660000   101.150000   
std       55.157145     0.726459      13.388989     6.650292     7.176719   

                              Joined           Contract Valid Until  \
count                           1627                           1627   
mean   2016-04-20 22:42:59.963122432  2021-05-11 04:26:24.265519360   
min              1991-06-01 00:00:00            2018-12-31 00:00:00   
25%              2015-07-01 00:00:00            2019-12-31 00:00:00   
50%              2017-06-27 00:00:00            2020-12-31 00:00:00   
75%              2018-07-01 00:00:00            2021-12-31 00:00:00   
max              2019-01-31 00:00:00            2026-12-31 00:00:00   
std                              NaN                            NaN   

       Release Clause (M_EUR)  
count             1627.000000  
mean                20.970138  
min                  0.000000  
25%                  3.850000  
50%                 12.800000  
75%                 26.650000  
max                228.100000  
std                 26.958951

Now our dataset finally looks clean and ready for the analysis.

## Step 6. Saving new dataset.

Last step is saving our dataset in '../data/processed/fifa_processed.csv'.

In [31]:
data.to_csv('../data/processed/fifa_processed.csv', index=False)